File: app/src/main/cpp/wn-steam-client/src/pb/cmsg_client_get_depot_decryption_key.cpp
#include "wn_steam/pb/cmsg_client_get_depot_decryption_key.h"

#include "wn_steam/proto_wire.h"

namespace wn_steam::pb {

std::vector<uint8_t> CMsgClientGetDepotDecryptionKey::serialize() const {
    std::vector<uint8_t> out;
    proto::Writer w(out);
    w.uint32_field(1, depot_id);
    w.uint32_field(2, app_id);
    return out;
}

std::optional<CMsgClientGetDepotDecryptionKeyResponse>
CMsgClientGetDepotDecryptionKeyResponse::deserialize(std::span<const uint8_t> body) noexcept {
    proto::Reader r(body);
    CMsgClientGetDepotDecryptionKeyResponse m;
    while (!r.eof()) {
        auto t = r.next_tag();
        if (!t) {
            if (!r.ok()) return std::nullopt;
            break;
        }
        switch (t->field_number) {
            case 1:
                if (auto v = r.u32(); v) m.eresult = *v;
                else return std::nullopt;
                break;
            case 2:
                if (auto v = r.u32(); v) m.depot_id = *v;
                else return std::nullopt;
                break;
            case 3: {
                auto b = r.bytes();
                if (!b) return std::nullopt;
                m.depot_encryption_key.assign(b->begin(), b->end());
                break;
            }
            default:
                if (!r.skip(t->wire_type)) return std::nullopt;
                break;
        }
    }
    return m;
}

}  // namespace wn_steam::pb

File: app/src/main/cpp/wn-steam-client/rust/src/pb/cmsg_client_get_depot_decryption_key.rs
use crate::proto_wire::{Reader, Writer};

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct CMsgClientGetDepotDecryptionKey {
    pub depot_id: u32,
    pub app_id: u32,
}

impl CMsgClientGetDepotDecryptionKey {
    pub fn serialize(&self) -> Vec<u8> {
        let mut out = Vec::new();
        let mut w = Writer::new(&mut out);
        w.uint32_field(1, self.depot_id);
        w.uint32_field(2, self.app_id);
        out
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct CMsgClientGetDepotDecryptionKeyResponse {
    pub eresult: u32,
    pub depot_id: u32,
    pub depot_encryption_key: Vec<u8>,
}

impl Default for CMsgClientGetDepotDecryptionKeyResponse {
    fn default() -> Self {
        Self {
            eresult: 2,
            depot_id: 0,
            depot_encryption_key: Vec::new(),
        }
    }
}

impl CMsgClientGetDepotDecryptionKeyResponse {
    pub fn deserialize(body: &[u8]) -> Option<Self> {
        let mut reader = Reader::new(body);
        let mut msg = Self::default();
        while !reader.eof() {
            let Some(tag) = reader.next_tag() else {
                return reader.ok().then_some(msg);
            };
            match tag.field_number {
                1 => msg.eresult = reader.u32()?,
                2 => msg.depot_id = reader.u32()?,
                3 => msg.depot_encryption_key = reader.bytes()?.to_vec(),
                _ => {
                    if !reader.skip(tag.wire_type) {
                        return None;
                    }
                }
            }
        }
        Some(msg)
    }
}
